From d37e3f77bdf61e1d1b3dc51332cbe8721eb430e1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=C3=86var=20Arnfj=C3=B6r=C3=B0=20Bjarmason?=
 <avarab@gmail.com>
Date: Thu, 16 May 2019 21:05:44 +0200
Subject: [RFC/PATCH 2/4] fsck: test for 2^63 overflowing in addition to 2^64
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Copy the test added in d4b8de0420 ("fsck: report integer overflow in
author timestamps", 2014-02-24) changing 18446744073709551617 to
9223372036854775809. This tests if a 64 bit signed timestamp_t and
time_t would overflow, in addition to the existing unsigned test.

The overflow check added in that change and amended in
7ca36d9398 ("date: check date overflow against time_t", 2014-02-24)
would already catch this, but it wasn't tested for.

On my system (and I believe all *nix) time_t is signed. POSIX just
says that it "shall be an integer type"[1].

1. https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_types.h.html

Signed-off-by: Ævar Arnfjörð Bjarmason <avarab@gmail.com>
---
 t/t1450-fsck.sh | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/t/t1450-fsck.sh b/t/t1450-fsck.sh
index 73ab0a8ee9..82435d01d2 100755
--- a/t/t1450-fsck.sh
+++ b/t/t1450-fsck.sh
@@ -210,6 +210,20 @@ test_expect_success 'integer overflow in timestamps is reported' '
 	test_i18ngrep "error in commit $new.*integer overflow" out
 '
 
+# date is 2^63 (timestamp_t is signed)
+test_expect_success 'signed integer overflow in timestamps is reported' '
+	git cat-file commit HEAD >basis &&
+	sed "s/^\\(author .*>\\) [0-9]*/\\1 9223372036854775808/" \
+		<basis >bad-timestamp &&
+	new=$(git hash-object -t commit -w --stdin <bad-timestamp) &&
+	test_when_finished "remove_object $new" &&
+	git update-ref refs/heads/bogus "$new" &&
+	test_when_finished "git update-ref -d refs/heads/bogus" &&
+	test_must_fail git fsck 2>out &&
+	cat out &&
+	test_i18ngrep "error in commit $new.*integer overflow" out
+'
+
 test_expect_success 'commit with NUL in header' '
 	git cat-file commit HEAD >basis &&
 	sed "s/author ./author Q/" <basis | q_to_nul >commit-NUL-header &&
-- 
2.21.0.1020.gf2820cf01a

